src/bootstrap/native.rs

```diff
@@ -465,6 +468,38 @@ impl Step for Llvm {
 
         cfg.build();
 
+        // When building LLVM with LLVM_LINK_LLVM_DYLIB for macOS, an unversioned
+        // libLLVM.dylib will be built. However, llvm-config will still look
+        // for a versioned path like libLLVM-14.dylib. Manually create a symbolic
+        // link to make llvm-config happy.
+        if llvm_link_shared && target.contains("apple-darwin") {
+            let mut cmd = Command::new(&build_llvm_config);
+            let version = output(cmd.arg("--version"));
+            let major = version.split('.').next().unwrap();
+            let lib_name = match llvm_suffix {
+                Some(s) => format!("lib/libLLVM-{}{}.dylib", major, s),
+                None => format!("lib/libLLVM-{}.dylib", major),
+            };
+
+            // The reason why we build the library path from llvm-config is because
+            // the output of llvm-config depends on its location in the file system.
+            // Make sure we create the symlink exactly where it's needed.
+            let llvm_base = build_llvm_config.parent().unwrap().parent().unwrap();
+            let lib_llvm = llvm_base.join(lib_name);
+            #[cfg(windows)]
+            if std::os::windows::fs::symlink_file("libLLVM.dylib", &lib_llvm).is_err() {
+                panic!("unable to create symbolic link {:?}", lib_llvm.file_name());
+            }
+            #[cfg(unix)]
+            if std::os::unix::fs::symlink("libLLVM.dylib", &lib_llvm).is_err() {
+                panic!("unable to create symbolic link {:?}", lib_llvm.file_name());
+            }
+            #[cfg(all(not(windows), not(unix)))]
+            {
+                panic!("unable to cross-compile target {}", target.triple);
+            }
```

Comment:
nit: this is the second place we symlink files, maybe add a `builder.symlink` helper? but this is fine too if it's a hassle

https://github.com/rust-lang/rust/blob/43347397f7c5ca9a670a3bb3890c7187e24a52ab/src/bootstrap/lib.rs#L1449

Reply:
Created a helper function as suggested